02-intermediate-ruby/references-and-values.md

```diff
@@ -0,0 +1,278 @@
+puts "After calling short_strings"
+puts "#{pets}"
+```
+
+The results:
+
+```
+Inside short_strings, input is
+["dog", "par", "cat", "lla"]
+After calling short_strings
+["dog", "parrot", "cat", "llama"]
+```
+
+`short_strings` is indeed creating a list of shortened words and storing it under the name `input`, but our outer variable `pets` isn't being updated. The reason why has to do with references and values, and how data is stored in a computer.
+
+As an aside: one way to fix our method is to return the new array, and when calling it say `pets = short_strings(pets)`. However, sometimes this isn't an option - for example, what if our method was supposed to return the number of words that had to be truncated?
+
+## References and Values
+
+When we create an array in Ruby (or a string or a hash or any other complex data type), we're actually creating two things.
+
+The first is the _value_ of the array, which involves asking the operating system for a bit of memory and then putting our data in it. You can think of this as the actual object. Each piece of memory we get from the OS has an _address_ representing its physical location in hardware, which is how we get back to it later.
+
+The second is a _reference_ to the array, which ties together the address of that memory with a name for our program to use. The address part of a reference is sometimes called a _pointer_ (especially in low-level languages like C and C++), and we say that a variable _points to_ or _references_ an object.
+
+Recalling our example above, saying
+
+```ruby
+pets = ['dog', 'parrot', 'cat', 'llama']
+```
+
+and imagining that the `pets` array has been stored at address `1234`, we would get the following memory layout:
+
+![references and variables](images/references_and_values/references_and_values.png)
+
+Every variable in Ruby consists of these two parts, a reference and a value. Normally when you type the variable's name, Ruby automatically goes and gets the object. If you want to find out what object your variable references, you can use the `object_id` method:
+
+```ruby
+pets = ["dog", "parrot", "cat", "llama"]
+puts "pets.object_id: #{pets.object_id}"
+
+# Different objects have different IDs
+veggies = ["turnip", "beet"]
+puts "veggies.object_id: #{veggies.object_id}"
+```
+
+### The `=` Operator
+
+The `=` operator changes what a variable points at.
+
+If we assign one variable to another variable, they will both reference the same underlying object.
+
+```ruby
+# Two variables can point to the same object
+repeat = veggies
+puts "repeat.object_id: #{repeat.object_id}" # same as veggie.object_id
+```
+
+![referencing a variable twice](images/references_and_values/two_pointers.png)
+
+If we make changes to the object through one variable, we can see the changes via the other. The variables have different names, but the underlying object is the same.
+
+```ruby
+veggies[1] = "onion"
+repeat.push("potato")
+puts "#{veggies}"     # ["turnip", "onion", "potato"]
```

Comment:
I think this code sample would be more meaningful if we moved this line above the `repeat.push` line and used the comment to show that its value before `push` was `["turnip", "onion"]`.

It'd probably also be a good idea to further highlight, in text after the code sample, that both times we're `puts`ing the variable `veggies`, but we called `push` on the variable `repeat`.